buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java

```diff
@@ -0,0 +1,125 @@
+
+    public PluginPropertiesTask() {
+        Project project = this.getProject();
+
+        File templateFile = new File(project.getBuildDir(), "templates/" + this.descriptorOutput.getName());
+
+        Task copyPluginPropertiesTemplate = project.getTasks().create("copyPluginPropertiesTemplate")
+                .doLast((unused) -> {
+                    InputStream resourceTemplate = PluginPropertiesTask.class.getResourceAsStream("/" + this.descriptorOutput.getName());
+                    templateFile.getParentFile().mkdirs();
+                    try {
+                        System.out.println(templateFile); ;
+                        Files.copy(resourceTemplate, templateFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+        dependsOn(copyPluginPropertiesTemplate);
+        extension = project.getExtensions().create("esplugin", PluginPropertiesExtension.class, project);
+        project.afterEvaluate((unused) -> {
+            // check require properties are set
+            if (this.extension.getName() == null) {
+                throw new InvalidUserDataException("name is a required setting for esplugin");
+            }
+            if (this.extension.getDescription() == null) {
+                throw new InvalidUserDataException("description is a required setting for esplugin");
+            }
+            if (this.extension.getClassname() == null) {
+                throw new InvalidUserDataException("classname is a required setting for esplugin");
+            }
+
+            from(templateFile.getParentFile()).include(this.descriptorOutput.getName());
+            into(this.descriptorOutput.getParentFile());
+            Map<String, String> properties = this.generateSubstitutions();
+            expand(properties);
+            this.getInputs().properties(properties);
+        });
+
+    }
+
+    private Map<String, String> generateSubstitutions() {
+        Map<String, String> substitutions = new HashMap<>();
+        substitutions.put("name", this.extension.getName());
+        substitutions.put("description", this.extension.getDescription());
+        substitutions.put("version", this.extension.getVersion());
+        substitutions.put("elasticsearchVersion", Version.fromString(VersionProperties.getElasticsearch()).toString());
+
+        //TODO (Raf): No getter for project.getTargetCompatibility()? What's the alternative option for below line?
+        substitutions.put("javaVersion", String.valueOf(this.getProject().getVersion()));
+        substitutions.put("classname", this.extension.getClassname());
+
+        //TODO (Raf): Should we worry about possible NPE on getExtendedPlugins() here?
+        substitutions.put("extendedPlugins", this.extension.getExtendedPlugins().stream().collect(Collectors.joining(",")));
```

Comment:
Will the `.getExtendedPlugins()` ever return `NULL`? If yes, then I will add a null-check otherwise the current `implementation` should suffice.

Reply:
I think it's fine, as long as we check it in the extension.